orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package filerepo
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	repoFilePermPrivateRW os.FileMode = 0600
+	repoDirPermPrivateRWX os.FileMode = 0700
+)
+
+type Repo struct {
+	fileRepoDir         string
+	fileSuffix          string
+	transientFileMarker string
+
+	mu sync.Mutex
+}
+
+type repoOption func(r *Repo)
+
+func WithTransientFileMarker(marker string) repoOption {
+	return func(r *Repo) {
+		r.transientFileMarker = marker
+	}
+}
+
+// New initializes a new file repo at repoParentDir/fileSuffix. It takes optional
+// repoOption arguments for modifying the default repo options. All file system operations
+// on the returned file repo are thread safe.
+func New(repoParentDir, fileSuffix string, opts ...repoOption) (*Repo, error) {
+	// default transientFileMarker to ~, can modify via opts
+	repo := &Repo{
+		transientFileMarker: "~",
+		fileSuffix:          fileSuffix,
+		fileRepoDir:         filepath.Join(repoParentDir, fileSuffix),
+	}
+
+	for _, opt := range opts {
+		opt(repo)
+	}
+
+	// Check that fileSuffix and transientFileMarker don't contain the os path separator
+	err := validateRepoOptions(repo.fileSuffix, repo.transientFileMarker)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make the file repo if it doesn't already exist
+	if err := os.MkdirAll(repo.fileRepoDir, repoDirPermPrivateRWX); err != nil {
+		return nil, err
+	}
+
+	files, err := ioutil.ReadDir(repo.fileRepoDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove existing transient files in the repo
+	for _, f := range files {
+		isTransientFile, err := filepath.Match("*"+repo.transientFileMarker, f.Name())
+		if err != nil {
+			return nil, err
+		}
+		if isTransientFile {
+			os.Remove(filepath.Join(repo.fileRepoDir, f.Name()))
+		}
+	}
```

Comment:
Either validate that the fileSuffix does not end with the transientFileMarker, or match against suffix+marker, rather than just the marker... otherwise this is going to remove legitimate files.